src/Sentry/Protocol/Span.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using Sentry.Internal.Extensions;
+
+namespace Sentry.Protocol
+{
+    // https://develop.sentry.dev/sdk/event-payloads/span
+    public class Span : ISpan, IJsonSerializable
+    {
+        public SentryId SpanId { get; }
+        public SentryId? ParentSpanId { get; }
+        public SentryId TraceId { get; set; }
+        public DateTimeOffset StartTimestamp { get; set; }
+        public DateTimeOffset EndTimestamp { get; set; }
+        public string Operation { get; }
+        public string? Description { get; set; }
+        public SpanStatus? Status { get; set; }
+        public bool IsSampled { get; set; }
+
+        private ConcurrentDictionary<string, string>? _tags;
+        public IReadOnlyDictionary<string, string> Tags => _tags ??= new ConcurrentDictionary<string, string>();
+
+        private ConcurrentDictionary<string, object>? _data;
+        public IReadOnlyDictionary<string, object> Data => _data ??= new ConcurrentDictionary<string, object>();
+
+        internal Span(SentryId? spanId = null, SentryId? parentSpanId = null, string operation = "unknown")
+        {
+            SpanId = spanId ?? SentryId.Create();
+            ParentSpanId = parentSpanId;
+            TraceId = SentryId.Create();
+            StartTimestamp = EndTimestamp = DateTimeOffset.Now;
+            Operation = operation;
+        }
+
+        public ISpan StartChild(string operation) => new Span(null, SpanId, operation);
+
+        public void Finish(SpanStatus status = SpanStatus.Ok)
+        {
+            EndTimestamp = DateTimeOffset.Now;
+            Status = status;
+        }
+
+        public void WriteTo(Utf8JsonWriter writer)
+        {
+            writer.WriteStartObject();
```

Comment:
Odd that this is a mutable object that will be accessed concurrently and we're reading its state here to serialize. We need to make sure that all members we're accessing can be accessed concurrently.

Reply:
Do you have a suggestion what we should do here? We could try "freezing" the span once it's finished, but it will be hard to guarantee that nothing gets written to the dictionaries.